Geometry/GEMGeometryBuilder/src/ME0GeometryParsFromDD.cc

```diff
@@ -8,75 +8,157 @@
+    doChambers = fv.nextSibling();
+  }  
+}
+
+void
+ME0GeometryParsFromDD::buildChamber(DDFilteredView& fv, ME0DetId detId, RecoIdealGeometry& rgeo)
+{
+  LogDebug("ME0GeometryParsFromDD") << "buildChamber "<<fv.logicalPart().name().name()
+				    <<" "<< detId <<std::endl;
+
+  std::vector<double> pars = getDimension(fv);
+  std::vector<double> vtra = getTranslation(fv);
+  std::vector<double> vrot = getRotation(fv);
+
+  rgeo.insert(detId.chamberId().rawId(), vtra, vrot, pars, {fv.logicalPart().name().name()});
+}
+
+void
+ME0GeometryParsFromDD::buildLayer(DDFilteredView& fv, ME0DetId detId, RecoIdealGeometry& rgeo)
 {
-  bool doSubDets = fview.firstChild();
-
-  while( doSubDets )
-  {
-    // Get the Base Muon Number
-    MuonDDDNumbering mdddnum( muonConstants );
-    MuonBaseNumber mbn = mdddnum.geoHistoryToBaseNumber( fview.geoHistory());
-
-    // Get the The ME0 det Id 
-    ME0NumberingScheme me0num( muonConstants );
-    ME0DetId rollDetId( me0num.baseNumberToUnitNumber( mbn ));
-
-    std::vector<double> dpar = fview.logicalPart().solid().parameters();
-    std::vector<double> pars;
-    pars.emplace_back( dpar[4]); // half bottom edge
-    pars.emplace_back( dpar[8]); // half top edge
-    pars.emplace_back( dpar[0]); // half apothem
-    pars.emplace_back( 0.4 ); // half thickness
-    pars.emplace_back( 0.0 ); // nStrips
-    pars.emplace_back( 0.0 ); // nPads
-
-    std::string name = fview.logicalPart().name().name();
-    std::vector<std::string> strpars;
-    strpars.emplace_back( name );
-
-    DDRotationMatrix rota = fview.rotation();
-
-    // ROOT::Math way
-    DD3Vector x, y, z;
-    rota.GetComponents( x, y, z );
-    std::vector<double> vrot(9);
-    vrot[0] = x.X();
-    vrot[1] = x.Y();
-    vrot[2] = x.Z();
-    vrot[3] = y.X();
-    vrot[4] = y.Y();
-    vrot[5] = y.Z();
-    vrot[6] = z.X();
-    vrot[7] = z.Y();
-    vrot[8] = z.Z();
- 
-    DDTranslation tran = fview.translation();
-    std::vector<double> vtra(3);
-    vtra[0] = tran.x();
-    vtra[1] = tran.y();
-    vtra[2] = tran.z();
-
-    rgeo.insert( rollDetId.rawId(), vtra, vrot, pars, strpars );
-    doSubDets = fview.nextSibling(); // go to next layer
+  LogDebug("ME0GeometryParsFromDD") << "buildLayer "<<fv.logicalPart().name().name()
+				    <<" "<< detId <<std::endl;
+
+  std::vector<double> pars = getDimension(fv);
+  std::vector<double> vtra = getTranslation(fv);
+  std::vector<double> vrot = getRotation(fv);
+
+  rgeo.insert(detId.layerId().rawId(), vtra, vrot, pars, {fv.logicalPart().name().name()});
+}
+
+void
+ME0GeometryParsFromDD::buildEtaPartition(DDFilteredView& fv, ME0DetId detId, RecoIdealGeometry& rgeo)
+{
+  LogDebug("ME0GeometryParsFromDD") << "buildEtaPartition "<<fv.logicalPart().name().name()
+				    <<" "<< detId <<std::endl;
+  
+  // EtaPartition specific parameter (nstrips and npads) 
+  DDValue numbOfStrips("nStrips");
+  DDValue numbOfPads("nPads");
+  std::vector<const DDsvalues_type* > specs(fv.specifics());
```

Comment:
could this be a const & ? eg  std::vector<const DDsvalues_type*> & specs = fv.specifics();